scripts/soroban-settings/SorobanSettingsUpgrade.py

```diff
@@ -0,0 +1,258 @@
+            send_transaction_data.hash)
+        if get_transaction_data.status != GetTransactionStatus.NOT_FOUND:
+            break
+        time.sleep(3)
+
+    print(f"transaction: {get_transaction_data}\n\n")
+
+    if get_transaction_data.status == GetTransactionStatus.SUCCESS:
+        assert get_transaction_data.result_meta_xdr is not None
+        transaction_meta = stellar_xdr.TransactionMeta.from_xdr(
+            get_transaction_data.result_meta_xdr
+        )
+        result = transaction_meta.v3.soroban_meta.return_value.address.contract_id.hash  # type: ignore
+    return result
+
+
+def upload_upgrade_bytes(contract_id, upgrade):
+    source = soroban_server.load_account(kp.public_key)
+
+    tx = (
+        TransactionBuilder(source, network_passphrase, base_fee=100)
+        .add_time_bounds(0, 0)
+        .append_invoke_contract_function_op(
+            contract_id=contract_id,
+            function_name="write",
+            parameters=[Bytes(upgrade.to_xdr_bytes())],
+        )
+        .build()
+    )
+
+    tx = soroban_server.prepare_transaction(tx)
+    tx.sign(kp)
+    send_transaction_data = soroban_server.send_transaction(tx)
+    print(f"sent transaction: {send_transaction_data}")
+
+    while True:
+        print("waiting for transaction to be confirmed...")
+        get_transaction_data = soroban_server.get_transaction(
+            send_transaction_data.hash)
+        if get_transaction_data.status != GetTransactionStatus.NOT_FOUND:
+            break
+        time.sleep(3)
+
+    print(f"transaction: {get_transaction_data}\n\n")
+
+    assert get_transaction_data.status == GetTransactionStatus.SUCCESS
+
+
+def get_upgrade_key(contract_id, upgrade_hash):
+    key = ConfigUpgradeSetKey(contract_id=Hash(
+        contract_id), content_hash=Hash(upgrade_hash))
+    return key
+
+
+"""
+The setup_upgrade mode will setup the ContractData required for the
+upgrade, and then print out the url encoded xdr to submit to core
+using the upgrades endpoint (upgrades?mode=set&configupgradesetkey=********).
+
+Note that the upgrade is specified in code at the moment (upgrade = get_upgrade_set())
+"""
+
+
+def setup_upgrade(args):
+    # Get upgrade xdr
+
+    upgrade = get_upgrade_set()
+    upgrade_xdr = upgrade.to_xdr()
+    upgrade_hash = utils.sha256(upgrade.to_xdr_bytes())
+
+    # Deploy contract
+    contract_id_hash = deploy_contract()
+    # print(f"contract id xdr: {Hash(contract_id_hash).to_xdr()}")
+    print(f"contract id hex: {Hash(contract_id_hash).to_xdr_bytes().hex()}")
+
+    contract_id_str = StrKey.encode_contract(contract_id_hash)
+    print(f"contract id: {contract_id_str}")
+
+    # Invoke contract
+    upload_upgrade_bytes(contract_id_str, upgrade)
+
+    # Get upgrade
+    upgrade_key = get_upgrade_key(contract_id_hash, upgrade_hash).to_xdr()
+    url_encoded_key = urllib.parse.quote(upgrade_key)
+    print(f"url encoded upgrade: {url_encoded_key}")
```

Comment:
Just to be sure, Core's server can handle the `/`s in the encoded key and won't treat them as paths? You can pass `safe=''` to [`quote`](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.quote) if not

Reply:
Yeah the `/` gets encoded.

Reply:
Yeah core handles it